core/desugarer.cpp

```diff
@@ -331,14 +345,38 @@ class Desugarer {
             if (!field.superSugar) continue;
             // We have to bind self/super from expr1 outside the class, as we copy the expression
             // into the field body.
-            AST *index = field.expr1;
             // Clone it so that we maintain the AST as a tree.
-            ClonePass(*alloc).expr(index);
+            AST *index = clone(field.expr1);
             // This will remove self/super.
             SubstituteSelfSuper(this, super_vars, counter).expr(index);
-            AST *super_f = make<SuperIndex>(field.expr1->location, EF, EF, index, EF, nullptr);
-            field.expr2 = make<Binary>(ast->location, EF, super_f, EF, BOP_PLUS, field.expr2);
-            field.superSugar = false;
+            if (true) {
```

Comment:
Yeah this was a temporary experiment.  For `e+: e'`, The true branch is

```
[e]: (if e in super then super[e] else std.zero(std.type(e')) + e'
```

The false branch is

```
[e]: if e in super then super[e] + e' else e'
```

This relates to our semantic discussion:  The two branches are equivalent in a world where `e in super` never fails.  So going forward I'll use the false branch because it's simpler.  The difference is that in the true branch, + is involved in both the two conditional cases.  For mixins `Inner = { g+: { x: 3 } }` and `Outer = { f+: Inner }` in the desugaring of Outer we have `e' = Inner`.

In the false branch this would mean Outer is desugared to `After = { f: if "f" in super then super.f + Inner else Inner }`.  The result of that desugaring is that `{ } + After` would be an error because `Inner` cannot stand by itself (the "g" in super throws an error).

In the true branch the result of `{ } + After` not Inner standing alone but `std.zero(std.type(Inner)) + Inner`, which is `{ } + Inner` which means by induction it will work for all nestings.

However all of this is academic now since we agreed on the permissive super :)

Regarding your description of your understanding, both versions work with strings.  `{} + ""` will do an implicit string conversion of the `{}` and indeed create a "{}" string rather than a HeapExtendedObject.  However `{} + ""` only occurs in a case like `{ f: {} } + {f+: ""}` which for both cases should yield `{ f: "{}" }`, which is correct.  In vm.cpp `countLeaves(self)` is only ever used when self has been taken from the stack, so it is always some sort of HeapObject  (A HeapString is not a HeapObject, as the base type of all garbage collected things is HeapEntity).  The else branch in countLeaves is just looking for other HeapObjects, i.e. HeapLeafObject (they have a common supertype).

Thanks for taking the time to figure out what you did, we really need more people to be confident in this code.